src/main/java/org/apache/maven/plugins/toolchain/jdk/ToolchainDiscoverer.java

```diff
@@ -0,0 +1,433 @@
+                        String val = f.getOrDefault(ENV, "");
+                        f.put(ENV, (val.isEmpty() ? "" : val + ",") + e.getKey());
+                    });
+
+            List<ToolchainModel> tcs = jdks.parallelStream()
+                    .map(s -> {
+                        ToolchainModel tc = getToolchainModel(s);
+                        flags.getOrDefault(s, Collections.emptyMap())
+                                .forEach((k, v) -> tc.getProvides().setProperty(k, v));
+                        String version = tc.getProvides().getProperty(VERSION);
+                        if (isLts(version)) {
+                            tc.getProvides().setProperty(LTS, "true");
+                        }
+                        return tc;
+                    })
+                    .sorted(getToolchainModelComparator(comparator))
+                    .collect(Collectors.toList());
+            if (this.cacheModified) {
+                writeCache();
+            }
+            PersistedToolchains ps = new PersistedToolchains();
+            ps.setToolchains(tcs);
+            return ps;
+        } catch (Exception e) {
+            if (log.isDebugEnabled()) {
+                log.warn("Error discovering toolchains: " + e, e);
+            } else {
+                log.warn("Error discovering toolchains (enable debug level for more information): " + e);
+            }
+            return new PersistedToolchains();
+        }
+    }
+
+    private static boolean isLts(String version) {
+        return version.startsWith("1.8.")
+                || version.startsWith("11.")
+                || version.startsWith("17.")
+                || version.startsWith("21.")
+                || version.startsWith("25.");
+    }
+
+    private void readCache() {
+        try {
+            cache = new ConcurrentHashMap<>();
+            Path cacheFile = getCacheFile();
+            if (Files.isRegularFile(cacheFile)) {
+                try (Reader r = Files.newBufferedReader(cacheFile)) {
+                    PersistedToolchains pt = new MavenToolchainsXpp3Reader().read(r, false);
+                    cache = pt.getToolchains().stream()
+                            // Remove stale entries
+                            .filter(tc -> {
+                                // If the bin/java executable is not available anymore, remove this TC
+                                if (!hasJavaC(getJdkHome(tc))) {
+                                    cacheModified = false;
+                                    return false;
+                                } else {
+                                    return true;
+                                }
+                            })
+                            .collect(Collectors.toConcurrentMap(this::getJdkHome, Function.identity()));
+                }
+            }
+        } catch (IOException | XmlPullParserException e) {
+            log.warn("Error reading toolchains cache: " + e);
```

Comment:
If the cache isn't read, what's the effect? if we just read directly without caching, then no warning should be logged. debug level at most.

How important is a cache here anyway? How long does it take to just recompute every time?

Reply:
The cache is imho quite important.  The length operation is the following: 
https://github.com/gnodet/maven-toolchains-plugin/blob/1cd938cfa39b262e669378135b7df89559408300/src/main/java/org/apache/maven/plugins/toolchain/jdk/ToolchainDiscoverer.java#L142-L154
The java process is spawned for each JDK, so on my computer, it takes 950 ms (using a the parallel fork join pool).  I'd rather avoid it each time the mojo is invoked.

But the cache is just a cache, if an error occurs while reading/writing, the process will still work, that's why errors are caught and logged at WARNING level.  I'm not sure to understand the "if we just read directly without caching" sentence.

Reply:
Note that the cache does not cache discovered toolchains (i.e. paths which contain a JDK), but rather the information related to each found JDK.